### PR TITLE
WIP: Tweak and fix a bunch of things

### PR DESCRIPTION
**Pull request by jepotter1-archive**
This is a really cool project! Despite being ostensibly just a demo game, it's already a lot better than most open source shooter games. And of course, it's written in Rust too 😛

This is a WIP document and I'll be adding more tasks/features as I think of them. Please let me know if you want me to fix/amend anything here.

Done:

- [x] Crouching is more pronounced and reduces movement speed.
- [x] The cosine wave is replaced with a more advanced function for view bobbing (more bumpy and uneven/asymmetrical for realism).
- [x] The player height is reduced.
- [x] The weapon offset is tweaked slightly to be straighter and more realistic.
- [x] Horizontal movement is disabled while jumping.
- [x] Bullet speed of AK47 and M16 is increased to a realistic level.
- [x] Ability to aim down sights with MOUSE2.
- [x] When you stand up from crouching it throws you into the air. The upwards force of this lurch seems to increase as you increase `stand_up_speed` and decrease `stand_body_height`. This seems to be a defect in the physics engine since I can't find any code that might do this inside this repository.
- [x] Ability to save settings to a file and load them on startup.
- [x] Weapon should lower when running and raise when crouching.
- [x] Reduce mouse sensitivity when ADS.
- [x] Move all the asset filename references to a centralized location for better maintainability.
- [x] Fix a few typos and grammatical errors in the README and put an image at the top.

Immediate to-dos:
- [ ] Framerate should be customizable.
- [ ] Screen should have effect when the player is damaged (blood marks, flash red, red ring perhaps?).
- [ ] Velocity seems to abruptly change and does not accelerate or decelerate like in real life. After you stop walking you should continue to slide just a little bit in the direction you were moving because inertia.
- [ ] When you hit the ground after jumping or falling the view should lurch downwards a bit (relative to how bad the fall was) because downwards inertia.

Need help (mainly physics related since I'm bad at physics):

- [ ] Staircases should be easier to climb and not cause a bunch of jittering.

Distant future to-dos (beyond my personal skill level, but cool to have):

- [ ] First and third person player models.
- [ ] Reloading with animations.
- [ ] Multiplayer networking with a client-server model.

### CODE REVIEW

**Comment by mrDIMAS** [comment]
Awesome, thanks! Ping me when you think its ready to be merged.

> Bullet speed of AK47 and M16 is increased to a realistic level (seriously, why was it so slow before?)

I was unsure if custom continuous collision detection for bullet can handle such speed. It is based on ray casting using position from previous frame. 

> When you stand up from crouching it throws you into the air.

Yes, this comes from physics engine. This is because player's body capsule increasing its height, but lower part of it becomes pushed into ground a bit and physics engine pushes it out. I fixed that in [d977a00
](https://github.com/mrDIMAS/rusty-shooter/commit/d977a008194481286bf9ceb3d9ac6a540e12b87d).

> Velocity seems to abruptly change and does not accelerate or decelerate like in real life.

This is because velocity in oXZ plane of player is controlled manually:

```rust
...
 if let Some(normalized_velocity) = velocity.normalized() {
            body.set_x_velocity(normalized_velocity.x * self.move_speed * speed_mult);
            body.set_z_velocity(normalized_velocity.z * self.move_speed * speed_mult);
            ...
}
...
```

To fix this you can introduce a new pair of variables, something like x_speed, z_speed and operate on them. Finally in update method you can do something like this:

```rust

...

body.set_x_velocity(self.x_speed);
body.set_z_velocity(self.z_speed);

const SLOWDOWN_SPEED: f32 = some_value;
self.x_speed = (self.x_speed - SLOWDOWN_SPEED).max(0.0);
self.z_speed = (self.z_speed - SLOWDOWN_SPEED).max(0.0);

```

> When you hit the ground after jumping or falling the view should lurch downwards a bit 

You can add new variable called something like head_landing_offset: f32, when player just touched the ground (but previousy was in air), you should set it to some value (-0.15?) and on each frame in update method do this:

```rust
const RECOVER_SPEED: f32 = some_value;
self.head_landing_offset= (self.head_landing_offset + RECOVER_SPEED).min(0.0);

...
  camera_node
            .local_transform_mut()
            .set_position(self.camera_offset + Vec3::new(0.0, self.head_landing_offset, 0.0));
```

> First and third person player models

For this we can use bot models, also there are lots of cool models (with animations) on mixamo.com 

> Reloading with animations

I think simple animations can be added in Blender pretty easily.

> Multiplayer networking with a client-server model

Most complex goal, probably will take few months to implement.

File: src/hud.rs [hunk 1]
```diff
@@ -76,24 +76,6 @@ impl Hud {
                 .with_width(frame_size.0 as f32)
                 .with_height(frame_size.1 as f32)
                 .with_visibility(false)
-                .with_child(
```
Comment (mrDIMAS) [inline src/hud.rs]: I'd put that in game options, something like "Show crosshair" instead of removing it completely.

Comment (jepotter1-archive) [reply src/hud.rs]: Ah okay. My thought was that the crosshair would be redundant with the laser.

Comment (mrDIMAS) [reply src/hud.rs]: Actually laser was just a raycasting test :) I thought to remove it but then decided to leave it,

**Comment by jepotter1-archive** [comment]
Abandoned because of too many merge conflicts.

**Pull request closed without merging**